include/seqan3/search/kmer_index/shape_iterator.hpp

```diff
@@ -0,0 +1,473 @@
+    using const_reference = shape_iterator const &;
+    //!\brief Tag this class depending on which concept it_t models.
+    using iterator_category = std::conditional_t<std::RandomAccessIterator<it_t>,
+                                                 typename std::random_access_iterator_tag,
+                                                 std::conditional_t<std::BidirectionalIterator<it_t>,
+                                                                    typename std::bidirectional_iterator_tag,
+                                                                    typename std::forward_iterator_tag>>;
+
+private:
+    //!\brief The size type of this iterator.
+    using position_type = std::make_unsigned_t<difference_type>;
+    //!\brief The alphabet type of the passed iterator.
+    using alphabet_t = value_type_t<it_t>;
+
+    //!\brief The hash value.
+    size_t hash_value{0};
+
+    //!\brief The seqan3::shape to use for hashing.
+    shape s;
+
+    //!\brief The factor for the left most position of the hash value.
+    size_t roll_factor{0};
+
+    //!\brief True if `s` contains only 1 (ungapped), false otherwise.
+    bool ungapped_shape{false};
+
+    //!\brief Iterator pointing to begin of the underlying text.
+    it_t text_start;
+
+    //!\brief Increments iterator by 1.
+    void hash_forward()
+    {
+        if (ungapped_shape)
+        {
+            hash_roll();
+        }
+        else
+        {
+            std::advance(text_left,  1);
+            hash_full();
+        }
+    }
+
+    /*!\brief Increments iterator by `skip`.
+     * \param skip Amount to increment.
+     * \attention This function is only avaible if it_t models std::RandomAccessIterator.
+     */
+    void hash_forward(difference_type const skip)
+        //!\cond
+        requires std::RandomAccessIterator<it_t>
+        //!\endcond
+    {
+        if (ungapped_shape && static_cast<size_t>(skip) < std::ranges::size(s))
+        {
+            for (difference_type i = 0; i < skip; ++i)
+                hash_roll();
+        }
+        else
+        {
+            std::advance(text_left,  skip);
+            hash_full();
+        }
+    }
+
+    /*!\brief Decrements iterator by 1.
+     * \attention This function is only avaible if it_t models std::BidirectionalIterator.
+     */
+    void hash_reverse()
+        //!\cond
+        requires std::BidirectionalIterator<it_t>
+        //!\endcond
+    {
+        if (ungapped_shape)
+        {
+            hash_roll_back();
+        }
+        else
+        {
+            std::advance(text_left,  -1);
+            hash_full();
+        }
+    }
+
+    /*!\brief Decrements iterator by `skip`.
+     * \param skip Amount to decrement.
+     * \attention This function is only avaible if it_t models std::BidirectionalIterator.
+     */
+    void hash_reverse(difference_type const skip)
+    {
+        if (ungapped_shape && static_cast<size_t>(skip) < std::ranges::size(s))
```

Comment:
why do you use ranges::size on a shape? (s is the shape I guess) it is not a range is it? and if you know the type and it has a member function you can use it I think.

Reply:
it is a `std::array`.. I'm not so sure about using the member function

Reply:
Anyway, the check is removed.